exporters/prometheus/exporter.go

```diff
@@ -233,3 +233,20 @@ func sanitizeRune(r rune) rune {
 	}
 	return '_'
 }
+
+// sanitizeName replaces characters which don't match the regex `^[a-zA-Z_:][a-zA-Z0-9_:]*$` with `_`.
+func sanitizeName(n string) string {
+	if len(n) == 0 {
+		return n
+	}
+	var sn = make([]string, len(n))
+	for i, b := range n {
+		if !((b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || b == '_' || b == ':' || (b >= '0' && b <= '9' && i > 0)) {
+			sn[i] = "_"
+		} else {
+			sn[i] = string(b)
+		}
+	}
+
+	return strings.Join(sn, "")
+}
```

Comment:
This allocates a slice even if the name is valid. I would suggest an alternate approach here, one based on `strings.Map`.


```go
func sanitizeName(n string) string {
	// This algorithm is based on strings.Map from Go 1.19.

	const (
		replacement = '_'
		width       = 1
	)

	valid := func(i int, r rune) bool {
		// Taken from
		// https://github.com/prometheus/common/blob/dfbc25bd00225c70aca0d94c3c4bb7744f28ace0/model/metric.go#L92-L102
		if (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || r == '_' || r == ':' || (r >= '0' && r <= '9' && i > 0) {
			return true
		}
		return false
	}

	// This output buffer b is initialized on demand, the first time a
	// character needs to be replaced.
	var b strings.Builder
	for i, c := range n {
		if valid(i, c) {
			continue
		}
		b.Grow(len(n))
		b.WriteString(n[:i])
		b.WriteRune(replacement)
		n = n[i+width:]
		break
	}

	// Fast path for unchanged input.
	if b.Cap() == 0 { // b.Grow was not called above.
		return n
	}

	for _, c := range n {
		// Due to inlining, it is more performant to invoke WriteByte rather
		// then WriteRune.
		if valid(1, c) { // We are guaranteed to not be at the start.
			b.WriteByte(byte(c))
		} else {
			b.WriteByte(byte(replacement))
		}
	}

	return b.String()
}
```

Reply:
Can you include memory allocations? That way we can see the impact it will have on GC and other memory handing processes.

Reply:
Here is benchstat with the original/latest versions:

```
name             old time/op    new time/op    delta
Collect1-12        17.1µs ± 0%    17.1µs ± 0%   ~     (p=1.000 n=1+1)
Collect10-12       34.5µs ± 0%    32.9µs ± 0%   ~     (p=1.000 n=1+1)
Collect100-12       214µs ± 0%     192µs ± 0%   ~     (p=1.000 n=1+1)
Collect1000-12     2.09ms ± 0%    1.82ms ± 0%   ~     (p=1.000 n=1+1)
Collect10000-12    22.3ms ± 0%    18.9ms ± 0%   ~     (p=1.000 n=1+1)

name             old alloc/op   new alloc/op   delta
Collect1-12        34.5kB ± 0%    34.4kB ± 0%   ~     (p=1.000 n=1+1)
Collect10-12       44.8kB ± 0%    43.7kB ± 0%   ~     (p=1.000 n=1+1)
Collect100-12       154kB ± 0%     142kB ± 0%   ~     (p=1.000 n=1+1)
Collect1000-12     1.33MB ± 0%    1.18MB ± 0%   ~     (p=1.000 n=1+1)
Collect10000-12    12.9MB ± 0%    11.2MB ± 0%   ~     (p=1.000 n=1+1)

name             old allocs/op  new allocs/op  delta
Collect1-12          44.0 ± 0%      37.0 ± 0%   ~     (p=1.000 n=1+1)
Collect10-12          293 ± 0%       223 ± 0%   ~     (p=1.000 n=1+1)
Collect100-12       2.84k ± 0%     2.05k ± 0%   ~     (p=1.000 n=1+1)
Collect1000-12      29.0k ± 0%     20.1k ± 0%   ~     (p=1.000 n=1+1)
Collect10000-12      299k ± 0%      201k ± 0%   ~     (p=1.000 n=1+1)
```